### PR TITLE
Replace the int ChannelId with a GUID/UUID ChannelId

### PR DESCRIPTION
**Pull request by vincentpierre**
Is this description really necessary if I am not targeting master ?

### Proposed change(s)

Replaced the int ChannelId with Guid.

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [x] Code refactor
- [x] Breaking change
- [x] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added updated the changelog (if applicable)
- [x] I have added necessary documentation (if applicable)
- [ ] I have updated the migration guide (if applicable)

### Other comments
Will do the documentation edits on `develop-sc-api` if this gets merged

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Grpc/RpcCommunicator.cs [hunk 1]
```diff
@@ -504,7 +504,7 @@ public static byte[] GetSideChannelMessage(Dictionary<int, SideChannel> sideChan
                         var messageList = sideChannel.MessageQueue;
                         foreach (var message in messageList)
                         {
-                            binaryWriter.Write(sideChannel.ChannelId);
+                            binaryWriter.Write(sideChannel.ChannelId.ToByteArray());
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Grpc/RpcCommunicator.cs]: was this part of the problem you were running into with GUID from C# --> Python?

File: com.unity.ml-agents/Runtime/SideChannel/SideChannel.cs [hunk 2]
```diff
@@ -1,46 +1,8 @@
 using System.Collections.Generic;
+using System;
 
 namespace MLAgents
 {
-
-    internal static class ReservedChannelIdCheck{
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/SideChannel/SideChannel.cs]: nice 👍

File: docs/Python-API.md [hunk 3]
```diff
@@ -324,12 +324,13 @@ the game :
 using UnityEngine;
 using MLAgents;
 using System.Text;
+using System;
 
 public class StringLogSideChannel : SideChannel
 {
     public StringLogSideChannel()
     {
-        ChannelId = (int)SideChannelType.UserSideChannelStart + 1;
```
Comment (surfnerd) [inline docs/Python-API.md]: much better

File: ml-agents-envs/mlagents_envs/environment.py [hunk 4]
```diff
@@ -443,13 +444,15 @@ def _flatten(cls, arr: Any) -> List[float]:
...
-        side_channels: Dict[int, SideChannel], data: bytes
+        side_channels: Dict[uuid.UUID, SideChannel], data: bytes
     ) -> None:
         offset = 0
         while offset < len(data):
             try:
-                channel_type, message_len = struct.unpack_from("<ii", data, offset)
-                offset = offset + 8
+                channel_type = uuid.UUID(bytes_le=bytes(data[offset : offset + 16]))
+                offset += 16
```
Comment (surfnerd) [inline ml-agents-envs/mlagents_envs/environment.py]: so 16 bytes on each side it seems. cool

File: ml-agents-envs/mlagents_envs/environment.py [hunk 5]
```diff
@@ -473,11 +476,14 @@ def _parse_side_channel_message(
                 )
 
     @staticmethod
-    def _generate_side_channel_data(side_channels: Dict[int, SideChannel]) -> bytearray:
+    def _generate_side_channel_data(
+        side_channels: Dict[uuid.UUID, SideChannel]
+    ) -> bytearray:
         result = bytearray()
         for channel_type, channel in side_channels.items():
```
Comment (surfnerd) [inline ml-agents-envs/mlagents_envs/environment.py]: nit: `channel_type` --> `channel_id`

File: com.unity.ml-agents/Runtime/Grpc/RpcCommunicator.cs (hunk 1 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Grpc/RpcCommunicator.cs]: Yes. I ended-up fixing on the python side

File: ml-agents-envs/mlagents_envs/environment.py (hunk 4 above)
Comment (vincentpierre) [reply ml-agents-envs/mlagents_envs/environment.py]: Yes, a GUID id 16 bytes or 2 long

**Pull request merged**